### PR TITLE
Add Add/Set/Remove (System) Zone cmdlets

### PR DESCRIPTION
**Pull request by CedricMoreau**

### CODE REVIEW

File: PowerFGT/Public/cmdb/system/interface.ps1 [hunk 1]
```diff
@@ -98,3 +98,73 @@ function Get-FGTSystemInterface {
     End {
     }
 }
+
+function Add-FGTSystemInterface {
+
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/interface.ps1:103]: Wrong PR

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 2]
```diff
@@ -95,6 +95,119 @@ function Get-FGTSystemZone {
...
+    )
+
+    Begin {
+    }
+
+    Process {
+
+        $zone = new-Object -TypeName PSObject
+        $zone | add-member -name "name" -membertype NoteProperty -Value $name
+        $zone | add-member -name "q_origin_key" -membertype NoteProperty -Value $name
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:135]: No needed

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 3]
```diff
@@ -95,6 +95,119 @@ function Get-FGTSystemZone {
...
+
+        if ( $PsBoundParameters.ContainsKey('intrazone') ) {
+            $zone | add-member -name "intrazone" -membertype NoteProperty -Value $intrazone
+        }
+
+        Invoke-FGTRestMethod -uri 'api/v2/cmdb/system/zone' -method 'POST' -body $zone -connection $connection | Out-Null
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:141]: Not possible to add member zone too ?

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 4]
```diff
@@ -95,6 +95,119 @@ function Get-FGTSystemZone {
...
+        Get-FGTSystemZone -name $name
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:142]: Add -connection parameter (for multi connection

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 5]
```diff
@@ -95,6 +95,119 @@ function Get-FGTSystemZone {
...
+        [psobject]$connection = $DefaultFGTConnection
+    )
+
+    Begin {
+    }
+
+    Process {
+
+        $zone = new-Object -TypeName PSObject
+        $get_interface = Get-FGTSystemInterface -name $interface
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:184]: ?

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 6]
```diff
@@ -95,6 +95,119 @@ function Get-FGTSystemZone {
...
+            }
+            $zone | add-member -name "interface" -membertype NoteProperty -Value $ports
+        }
+
+        if ( $PsBoundParameters.ContainsKey('intrazone') ) {
+            $zone | add-member -name "intrazone" -membertype NoteProperty -Value $intrazone
+        }
+
+        Invoke-FGTRestMethod -uri "api/v2/cmdb/system/zone/${name}" -method 'PUT' -body $zone -connection $connection | Out-Null
+        Get-FGTSystemZone -name $name
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:207]: -connection

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 7]
```diff
@@ -95,6 +95,119 @@ function Get-FGTSystemZone {
...
+        {
+            Throw "The interface specified does not exist"
+        }
+
+        if ( $PsBoundParameters.ContainsKey('interfaces') ) {
+            $ports = @()
+            foreach ( $member in $interfaces ) {
+                $member_attributes = @{}
+                $member_attributes.add( 'interface-name', $member)
+                $member_attributes.add( 'q_origin_key', $member)
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:196]: No needed

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 8]
```diff
@@ -131,15 +133,42 @@ function Add-FGTSystemZone {
     Process {
 
         $zone = new-Object -TypeName PSObject
-        $zone | add-member -name "name" -membertype NoteProperty -Value $name
-        $zone | add-member -name "q_origin_key" -membertype NoteProperty -Value $name
+        if ( $PsBoundParameters.ContainsKey('name') ) {
+            $zone | add-member -name "name" -membertype NoteProperty -Value $name
+        }
+        Else{
+            Throw "You need to specifiy a name"
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:140]: Use mandatory parameter

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 9]
```diff
@@ -95,6 +95,190 @@ function Get-FGTSystemZone {
...
+
+        If($null -ne $get_zone_name)
+        {
+            Throw "Already a zone using the same name"
+        }
+
+        if ( $PsBoundParameters.ContainsKey('interfaces') ) {
+            $ports = @()
+            foreach ( $member in $interfaces ) {
+                $get_interface = Get-FGTSystemInterface -name $member
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:153]: Missing connection parameter and don’t check for the moment, I will plan a cache for avoid to check for each interface (can be very slow...)

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 10]
```diff
@@ -95,6 +95,190 @@ function Get-FGTSystemZone {
...
+        .EXAMPLE
+        Set-FGTSystemZone -name PowerFGT -intrazone deny -interfaces port5,port6
+
+        Set the zone named PowerFGT with intra-zone traffic authorized, and with port 5 and port 6 bound to it
+    #>
+
+    [CmdletBinding(DefaultParameterSetName = "default")]
+    Param(
+        [Parameter (Mandatory = $true, Position = 1)]
+        [string]$name,
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:197]: Use pipeline (And use another variable like zone)

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 11]
```diff
@@ -95,6 +95,190 @@ function Get-FGTSystemZone {
...
+        .EXAMPLE
+        Remove-FGTSystemZone -name PowerFGT
+
+        Remove the zone named PowerFGT
+    #>
+
+    [CmdletBinding(DefaultParameterSetName = "default")]
+    Param(
+        [Parameter (Mandatory = $true, ValueFromPipeline = $true, Position = 1)]
+        [string]$name,
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:268]: Add validate zone

File: Tests/integration/Zone.Tests.ps1 [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+# SPDX-License-Identifier: Apache-2.0
+#
+
+#include common configuration
+. ../common.ps1
+
+Describe "Get zone" {
+
+    BeforeAll {
+        Add-FGTSystemZone -name "PowerFGT" -intrazone deny -interfaces port9,port10
```
Comment (alagoutte) [inline Tests/integration/Zone.Tests.ps1:13]: Use variable for name of zone

File: Tests/integration/Zone.Tests.ps1 [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    It "Try to add zone PowerFGT (but there is already an object with same name)" {
+        #Add first zone
+        Add-FGTSystemZone -name PowerFGT
+        #Add Second zone with same name
+        { Add-FGTSystemZone -name PowerFGT } | Should -Throw "Already a zone using the same name"
+    }
+
+    It "Try to add zone without name" {
```
Comment (alagoutte) [inline Tests/integration/Zone.Tests.ps1:74]: Not needed

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 14]
```diff
@@ -95,6 +95,190 @@ function Get-FGTSystemZone {
...
+        [Parameter(Mandatory = $false)]
+        [psobject]$connection = $DefaultFGTConnection
+    )
+
+    Begin {
+    }
+
+    Process {
+
+        $null = Invoke-FGTRestMethod -method "DELETE" -uri "api/v2/cmdb/system/zone/${name}" -connection $connection
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:278]: Add should procees

File: Tests/integration/Zone.Tests.ps1 [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+}
+
+Describe "Add zone" {
+
+    AfterEach {
+        Remove-FGTSystemZone -name PowerFGT
+    }
+
+    It "Add zone PowerFGT" {
```
Comment (alagoutte) [inline Tests/integration/Zone.Tests.ps1:58]: Try with intrazone deny or allow

Try with 0 interface, 1, 2 ...

File: PowerFGT/Private/Confirm.ps1 [hunk 16]
```diff
@@ -128,4 +128,27 @@ Function Confirm-FGTVip {
...
+
+    Param (
+        [Parameter (Mandatory = $true)]
+        [object]$argument
+    )
+
+    #Check if it looks like an Zone element
+
+    if ( -not ( $argument | get-member -name name -Membertype Properties)) {
+        throw "Element specified does not contain an name property."
```
Comment (alagoutte) [inline PowerFGT/Private/Confirm.ps1:143]: an -> a name

File: PowerFGT/Private/Confirm.ps1 [hunk 17]
```diff
@@ -128,4 +128,27 @@ Function Confirm-FGTVip {
...
+    }
+    if ( -not ( $argument | get-member -name intrazone -Membertype Properties)) {
+        throw "Element specified does not contain a intrazone property."
```
Comment (alagoutte) [inline PowerFGT/Private/Confirm.ps1:146]: a -> an intrazone

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 18]
```diff
@@ -95,6 +95,300 @@ function Get-FGTSystemZone {
...
+                $ports += $member_attributes
+            }
+            $zone | add-member -name "interface" -membertype NoteProperty -Value $ports
+        }
+
+        if ( $PsBoundParameters.ContainsKey('intrazone') ) {
+            $zone | add-member -name "intrazone" -membertype NoteProperty -Value $intrazone
+        }
+
+        Invoke-FGTRestMethod -uri 'api/v2/cmdb/system/zone' -method 'POST' -body $zone -connection $connection | Out-Null
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:158]: Add variable for uri ?

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 19]
```diff
@@ -95,6 +95,300 @@ function Get-FGTSystemZone {
...
+}
+
+function Set-FGTSystemZone {
+
+    <#
+        .SYNOPSIS
+        Set a zone
+
+        .DESCRIPTION
+        Set a zone
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:174]: Configure a zone (name, intrazone, member...)

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 20]
```diff
@@ -95,6 +95,300 @@ function Get-FGTSystemZone {
...
+
+        .EXAMPLE
+        Set-FGTSystemZone -name PowerFGT -intrazone deny -interfaces port5,port6
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:177]: add more example (only change name, intrazone, member...)

Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:177]: also use pipeline for example

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 21]
```diff
@@ -95,6 +95,300 @@ function Get-FGTSystemZone {
...
+        Set the zone named PowerFGT with intra-zone traffic authorized, and with port 5 and port 6 bound to it
+    #>
+
+    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'medium')]
+    Param(
+        [Parameter (Mandatory = $true, ValueFromPipeline = $true, Position = 1)]
+        [ValidateScript({Confirm-FGTZone $_})]
+        [psobject]$zone,
+        [Parameter (Mandatory = $false)]
+        [string]$zone_name,
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:188]: name ?

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 22]
```diff
@@ -95,6 +95,300 @@ function Get-FGTSystemZone {
...
+
+    <#
+        .SYNOPSIS
+        Remove a zone
+
+        .DESCRIPTION
+        Remove a zone
+
+        .EXAMPLE
+        Remove-FGTSystemZone -zone PowerFGT
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:249]: add example with confirm parameter

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 23]
```diff
@@ -95,6 +95,300 @@ function Get-FGTSystemZone {
...
+
+        Remove the zone named PowerFGT
+    #>
+
+    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'medium')]
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:254]: high impact !

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 24]
```diff
@@ -95,6 +95,300 @@ function Get-FGTSystemZone {
...
+
+    <#
+        .SYNOPSIS
+        Remove a zone member
+
+        .DESCRIPTION
+        Remove a zone member
+
+        .EXAMPLE
+        Remove-FGTSystemZoneMember -name PowerFGT -interface port9
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:288]: a example with multiple port ?

Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:288]: use also pipeline for example

File: PowerFGT/Public/cmdb/system/zone.ps1 [hunk 25]
```diff
@@ -95,6 +95,300 @@ function Get-FGTSystemZone {
...
+
+    <#
+        .SYNOPSIS
+        Add a zone member
+
+        .DESCRIPTION
+        Add a zone member
+
+        .EXAMPLE
+        Add-FGTSystemZoneMember -name PowerFGT -interface port9
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/system/zone.ps1:350]: pipeline for eample...

File: Tests/common.ps1 [hunk 26]
```diff
@@ -16,6 +16,12 @@ $script:pester_vip1 = "pester_vip1"
 $script:pester_vip2 = "pester_vip2"
 $script:pester_policy1 = "pester_policy1"
 $script:pester_policy2 = "pester_policy2"
+$script:pester_port1 = "port7"
+$script:pester_port2 = "port8"
+$script:pester_port3 = "port9"
+$script:pester_port4 = "port10"
+$script:pester_zone1 = "PowerFGT1"
```
Comment (alagoutte) [inline Tests/common.ps1:23]: pester_zone1 ? (same after)

File: Tests/integration/Zone.Tests.ps1 [hunk 27]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    It "Change intrazone in deny" {
+        Get-FGTSystemZone -name $pester_zone1 | Set-FGTSystemZone -intrazone deny
+        $zone = Get-FGTSystemZone -name $pester_zone1
+        $zone.intrazone | Should -Be "deny"
+    }
+
+    It "Change interfaces" {
+        Get-FGTSystemZone -name $pester_zone1 | Set-FGTSystemZone -interfaces $pester_port1,$pester_port2
```
Comment (alagoutte) [inline Tests/integration/Zone.Tests.ps1:137]: it is the same interface (use $pester_port3 and 4?)

File: Tests/integration/Zone.Tests.ps1 [hunk 28]
```diff
@@ -0,0 +1,225 @@
...
+
+Describe "Remove zone" {
+
+    BeforeEach {
+        Add-FGTSystemZone -name $pester_zone1
+    }
+
+    It "Remove zone $pester_zone1 by filtering the name" {
+        $zone = Get-FGTSystemZone -name $pester_zone1
+        Get-FGTSystemZone $pester_zone1 | Remove-FGTSystemZone
```
Comment (alagoutte) [inline Tests/integration/Zone.Tests.ps1:158]: it is the same test ?

**Review by alagoutte — CHANGES_REQUESTED** [review]
Look good but need some change before merge

File: Tests/integration/Zone.Tests.ps1 [hunk 29]
```diff
@@ -0,0 +1,225 @@
+#
+# Copyright 2020, Alexis La Goutte <alexis dot lagoutte at gmail dot com>
+#
```
Comment (alagoutte) [inline Tests/integration/Zone.Tests.ps1:3]: add your credit (and also on Zone.ps1 ;))

**Comment by jelmerj** [comment]
This pull request needs support for the vdom parameter. Please note that the vdom parameter also needs to be added to the Get-FGT* calls which are called from within the functions

**Comment by alagoutte** [comment]
> 
> 
> This pull request needs support for the vdom parameter. Please note that the vdom parameter also needs to be added to the Get-FGT* calls which are called from within the functions
> I could help to add it, but I'm not an expert on Git and don't know how I could contribute to a pull request from someone else.

Hi, Good catch thanks !
it is missing for this cmdlet ! (i need to look for add test with vdom...)

What other get cmdlet do you have where missing vdom parameter ?

i can help you about git, you can contact me by mail (look file header) or other (twitter...)

**Pull request merged**